File: test/models/memory_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:xiaoyu_memory/models/memory.dart';

void main() {
  group('Memory Model', () {
    test('fromJson creates memory with all fields', () {
      final json = {
        'id': 'mem-123',
        'user_id': 'user-1',
        'type': 'text',
        'content': '测试内容',
        'summary': '测试总结',
        'title': '测试标题',
        'tags': ['tag1', 'tag2'],
        'media_url': 'https://example.com/image.jpg',
        'created_at': '2026-01-01T00:00:00.000Z',
        'updated_at': '2026-01-01T00:00:00.000Z',
      };
      final memory = Memory.fromJson(json);

      expect(memory.id, 'mem-123');
      expect(memory.userId, 'user-1');
      expect(memory.type, 'text');
      expect(memory.content, '测试内容');
      expect(memory.summary, '测试总结');
      expect(memory.title, '测试标题');
      expect(memory.tags.length, 2);
      expect(memory.mediaUrl, 'https://example.com/image.jpg');
    });

    test('fromJson handles missing fields with defaults', () {
      final json = {'id': 'mem-456', 'user_id': 'user-2', 'type': 'text', 'content': '', 'created_at': '', 'updated_at': ''};
      final memory = Memory.fromJson(json);

      expect(memory.id, 'mem-456');
      expect(memory.type, 'text');
      expect(memory.summary, '');
      expect(memory.title, '');
      expect(memory.tags, isEmpty);
      expect(memory.mediaUrl, isNull);
    });

    test('toJson serializes correctly', () {
      final memory = Memory(
        id: 'mem-789',
        userId: 'user-3',
        type: 'voice',
        content: '语音内容',
        summary: '总结',
        title: '标题',
        tags: ['tagA'],
        mediaUrl: null,
        createdAt: DateTime.utc(2026, 1, 1),
        updatedAt: DateTime.utc(2026, 1, 1),
      );
      final json = memory.toJson();

      expect(json['id'], 'mem-789');
      expect(json['user_id'], 'user-3');
      expect(json['type'], 'voice');
      expect(json['content'], '语音内容');
      expect(json['summary'], '总结');
      expect(json['tags'], ['tagA']);
    });

    test('copyWith creates new instance with updated fields', () {
      final original = Memory(
        id: 'mem-original',
        userId: 'user-1',
        type: 'text',
        content: '原始内容',
        createdAt: DateTime.utc(2026, 1, 1),
        updatedAt: DateTime.utc(2026, 1, 1),
      );
      final updated = original.copyWith(content: '新内容', title: '新标题');

      expect(updated.id, 'mem-original');
      expect(updated.content, '新内容');
      expect(updated.title, '新标题');
      expect(original.content, '原始内容'); // original unchanged
    });
  });
}

File: backend/internal/model/memory_test.go
package model

import (
	"testing"
	"time"
)

func TestMemory_ToJSON(t *testing.T) {
	now := time.Now()
	m := Memory{
		ID:        "mem-123",
		UserID:    "user-1",
		Type:      "text",
		Content:   "测试内容",
		Summary:   "测试总结",
		Title:     "测试标题",
		Tags:      []string{"tag1", "tag2"},
		MediaURL:  "https://example.com/image.jpg",
		CreatedAt: now,
		UpdatedAt: now,
	}

	json := m.ToJSON()

	if json["id"] != "mem-123" {
		t.Errorf("expected id 'mem-123', got %v", json["id"])
	}
	if json["user_id"] != "user-1" {
		t.Errorf("expected user_id 'user-1', got %v", json["user_id"])
	}
	if json["type"] != "text" {
		t.Errorf("expected type 'text', got %v", json["type"])
	}
	if json["content"] != "测试内容" {
		t.Errorf("expected content '测试内容', got %v", json["content"])
	}
	if len(json["tags"].([]string)) != 2 {
		t.Errorf("expected 2 tags, got %v", len(json["tags"].([]string)))
	}
}

func TestMemory_JSONMap(t *testing.T) {
	json := map[string]interface{}{
		"id":         "mem-456",
		"user_id":    "user-2",
		"type":       "voice",
		"content":    "语音内容",
		"summary":    "总结",
		"title":      "标题",
		"tags":       []string{"tagA"},
		"media_url":  "",
		"created_at": time.Now(),
		"updated_at": time.Now(),
	}

	m := JSONToMemory(json)

	if m.ID != "mem-456" {
		t.Errorf("expected ID 'mem-456', got %v", m.ID)
	}
	if m.Type != "voice" {
		t.Errorf("expected Type 'voice', got %v", m.Type)
	}
	if m.Content != "语音内容" {
		t.Errorf("expected Content '语音内容', got %v", m.Content)
	}
}

func TestUser_ToJSON(t *testing.T) {
	now := time.Now()
	u := User{
		ID:        "user-123",
		Phone:     "13800138000",
		Email:     "test@example.com",
		CreatedAt: now,
	}

	json := u.ToJSON()

	if json["id"] != "user-123" {
		t.Errorf("expected id 'user-123', got %v", json["id"])
	}
	if json["phone"] != "13800138000" {
		t.Errorf("expected phone '13800138000', got %v", json["phone"])
	}
	if json["email"] != "test@example.com" {
		t.Errorf("expected email 'test@example.com', got %v", json["email"])
	}
}
